From 20dc518659eda586b3e467ed43ad7c2d767e82cc Mon Sep 17 00:00:00 2001
From: William Bright <William.Bright@imd-tec.com>
Date: Wed, 8 Apr 2026 10:43:05 +0100
Subject: [PATCH 3/5] media: i2c: ar1335: Add ON Semiconductor AR1335 camera
 sensor driver

Add a V4L2 subdevice driver for the ON Semiconductor AR1335 13MP CMOS
image sensor. The driver supports up to 4K video capture over MIPI CSI-2
with 4 data lanes, and provides controls for exposure and gain.

Signed-off-by: William Bright <william.bright@imd-tec.com>
Upstream-Status: Inappropriate [Needs checking with the
v4l2-compliance tool]
---
 drivers/media/i2c/Kconfig  |    9 +
 drivers/media/i2c/Makefile |    1 +
 drivers/media/i2c/ar1335.c | 1017 ++++++++++++++++++++++++++++++++++++
 3 files changed, 1027 insertions(+)
 create mode 100644 drivers/media/i2c/ar1335.c

diff --git a/drivers/media/i2c/Kconfig b/drivers/media/i2c/Kconfig
index 5eb1e0e0a87a..1be1900f6139 100644
--- a/drivers/media/i2c/Kconfig
+++ b/drivers/media/i2c/Kconfig
@@ -60,6 +60,15 @@ config VIDEO_AR0521
 	  To compile this driver as a module, choose M here: the
 	  module will be called ar0521.
 
+config VIDEO_AR1335
+	tristate "ON Semiconductor AR1335 sensor support"
+	help
+	  This is a Video4Linux2 sensor driver for the ON Semiconductor
+	  AR1335 camera.
+
+	  To compile this driver as a module, choose M here: the
+	  module will be called ar1335.
+
 config VIDEO_GC0308
 	tristate "GalaxyCore GC0308 sensor support"
 	select V4L2_CCI_I2C
diff --git a/drivers/media/i2c/Makefile b/drivers/media/i2c/Makefile
index a3a6396df3c4..f1abc1bad7f2 100644
--- a/drivers/media/i2c/Makefile
+++ b/drivers/media/i2c/Makefile
@@ -16,6 +16,7 @@ obj-$(CONFIG_VIDEO_ADV7511) += adv7511-v4l2.o
 obj-$(CONFIG_VIDEO_ADV7604) += adv7604.o
 obj-$(CONFIG_VIDEO_ADV7842) += adv7842.o
 obj-$(CONFIG_VIDEO_AK7375) += ak7375.o
+obj-$(CONFIG_VIDEO_AR1335) += ar1335.o
 obj-$(CONFIG_VIDEO_AK881X) += ak881x.o
 obj-$(CONFIG_VIDEO_ALVIUM_CSI2) += alvium-csi2.o
 obj-$(CONFIG_VIDEO_APTINA_PLL) += aptina-pll.o
diff --git a/drivers/media/i2c/ar1335.c b/drivers/media/i2c/ar1335.c
new file mode 100644
index 000000000000..5fb645382301
--- /dev/null
+++ b/drivers/media/i2c/ar1335.c
@@ -0,0 +1,1017 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * ON Semiconductor AR1335 CMOS Image Sensor driver
+ *
+ * Copyright (C) 2026 IMD Technologies Ltd
+ */
+
+#include <linux/clk.h>
+#include <linux/delay.h>
+#include <linux/gpio/consumer.h>
+#include <linux/i2c.h>
+#include <linux/pm_runtime.h>
+#include <linux/regulator/consumer.h>
+#include <linux/videodev2.h>
+#include <media/v4l2-ctrls.h>
+#include <media/v4l2-device.h>
+#include <media/v4l2-fwnode.h>
+#include <media/v4l2-subdev.h>
+
+#define AR1335_NAME			"ar1335"
+
+#define AR1335_CHIP_ID			0x0153
+#define AR1335_NUM_SUPPLIES		3
+
+/* External clock (extclk) frequencies */
+#define AR1335_EXTCLK_MIN		(6 * 1000 * 1000)
+#define AR1335_EXTCLK_MAX		(48 * 1000 * 1000)
+
+/* PLL limits */
+#define AR1335_PLL_MIN			(320 * 1000 * 1000)
+#define AR1335_PLL_MAX			(1200 * 1000 * 1000)
+
+/* Pixel clock */
+#define AR1335_PIXEL_CLOCK_RATE		(220 * 1000 * 1000)
+#define AR1335_PIXEL_CLOCK_MIN		(168 * 1000 * 1000)
+#define AR1335_PIXEL_CLOCK_MAX		(414 * 1000 * 1000)
+
+/* Sensor pixel array limits */
+#define AR1335_MIN_X_ADDR_START		8u
+#define AR1335_MIN_Y_ADDR_START		8u
+#define AR1335_MAX_X_ADDR_END		4231u
+#define AR1335_MAX_Y_ADDR_END		3143u
+
+#define AR1335_WIDTH_MIN		8u
+#define AR1335_WIDTH_MAX		4224u
+#define AR1335_HEIGHT_MIN		8u
+#define AR1335_HEIGHT_MAX		3136u
+
+#define AR1335_WIDTH_BLANKING_MIN	240u
+#define AR1335_HEIGHT_BLANKING_MIN	142u
+#define AR1335_TOTAL_HEIGHT_MAX		65535u
+#define AR1335_TOTAL_WIDTH_MAX		65532u
+
+/* Analog gain */
+#define AR1335_ANA_GAIN_MIN		0x00
+#define AR1335_ANA_GAIN_MAX		0x3f
+#define AR1335_ANA_GAIN_STEP		0x01
+#define AR1335_ANA_GAIN_DEFAULT		0x00
+
+/* AR1335 registers */
+#define AR1335_REG_CHIP_ID		0x0000
+#define AR1335_REG_CCP_DATA_FORMAT	0x0112
+#define AR1335_REG_VT_PIX_CLK_DIV	0x0300
+#define AR1335_REG_FRAME_LENGTH_LINES	0x0340
+#define AR1335_REG_COARSE_INTEGRATION_TIME 0x3012
+#define AR1335_REG_ROW_SPEED		0x3016
+#define AR1335_REG_RESET		0x301A
+#define AR1335_REG_RESET_DEFAULTS	0x0238
+#define AR1335_REG_RESET_STREAM		BIT(2)
+#define AR1335_REG_ANA_GAIN_CODE_GLOBAL	0x3028
+#define AR1335_REG_GREEN1_GAIN		0x3056
+#define AR1335_REG_BLUE_GAIN		0x3058
+#define AR1335_REG_RED_GAIN		0x305A
+#define AR1335_REG_GREEN2_GAIN		0x305C
+#define AR1335_REG_GLOBAL_GAIN		0x305E
+#define AR1335_REG_HISPI_TEST_MODE	0x3066
+#define AR1335_REG_HISPI_TEST_MODE_LP11	0x0004
+#define AR1335_REG_TEST_PATTERN_MODE	0x3070
+#define AR1335_REG_FRAME_PREAMBLE	0x31B0
+#define AR1335_REG_LINE_PREAMBLE	0x31B2
+#define AR1335_REG_MIPI_TIMING_0	0x31B4
+#define AR1335_REG_MIPI_TIMING_1	0x31B6
+#define AR1335_REG_MIPI_TIMING_2	0x31B8
+#define AR1335_REG_MIPI_TIMING_3	0x31BA
+#define AR1335_REG_MIPI_TIMING_4	0x31BC
+#define AR1335_REG_SERIAL_FORMAT	0x31AE
+#define AR1335_REG_SERIAL_FORMAT_MIPI	0x0200
+#define AR1335_REG_HISPI_CONTROL_STATUS	0x31C6
+
+struct ar1335_reg {
+	u16 addr;
+	u16 val;
+};
+
+static const struct ar1335_reg ar1335_init_regs[] = {
+	/* MIPI timing */
+	{ AR1335_REG_FRAME_PREAMBLE,	0x005C },
+	{ AR1335_REG_LINE_PREAMBLE,	0x002D },
+	{ AR1335_REG_MIPI_TIMING_0,	0x2412 },
+	{ AR1335_REG_MIPI_TIMING_1,	0x142A },
+	{ AR1335_REG_MIPI_TIMING_2,	0x2413 },
+	{ AR1335_REG_MIPI_TIMING_3,	0x1C70 },
+	{ AR1335_REG_MIPI_TIMING_4,	0x868B },
+	/* 10-bit Bayer output */
+	{ AR1335_REG_CCP_DATA_FORMAT,	0x0A0A },
+};
+
+static const char * const ar1335_supply_names[] = {
+	"vdd_io",
+	"vdd",
+	"vaa",
+};
+
+/* link_freq = pixel_clock * bpp / (2 * num_lanes) = 220MHz * 10 / (2 * 4) */
+static const s64 ar1335_link_frequencies[] = {
+	275000000,
+};
+
+struct ar1335_ctrls {
+	struct v4l2_ctrl_handler handler;
+	struct {
+		struct v4l2_ctrl *gain;
+		struct v4l2_ctrl *red_balance;
+		struct v4l2_ctrl *blue_balance;
+	};
+	struct {
+		struct v4l2_ctrl *hblank;
+		struct v4l2_ctrl *vblank;
+	};
+	struct v4l2_ctrl *pixrate;
+	struct v4l2_ctrl *exposure;
+	struct v4l2_ctrl *test_pattern;
+};
+
+struct ar1335_dev {
+	struct i2c_client *i2c_client;
+	struct v4l2_subdev sd;
+	struct media_pad pad;
+	struct clk *extclk;
+	u32 extclk_freq;
+
+	struct regulator_bulk_data supplies[AR1335_NUM_SUPPLIES];
+	struct gpio_desc *reset_gpio;
+
+	/* lock to protect all members below */
+	struct mutex lock;
+	struct v4l2_mbus_framefmt fmt;
+	struct ar1335_ctrls ctrls;
+	unsigned int lane_count;
+	struct {
+		u16 pre;
+		u16 mult;
+		u16 pre2;
+		u16 mult2;
+		u16 vt_pix;
+	} pll;
+};
+
+static inline struct ar1335_dev *to_ar1335_dev(struct v4l2_subdev *sd)
+{
+	return container_of(sd, struct ar1335_dev, sd);
+}
+
+static inline struct v4l2_subdev *ctrl_to_sd(struct v4l2_ctrl *ctrl)
+{
+	return &container_of(ctrl->handler, struct ar1335_dev,
+			     ctrls.handler)->sd;
+}
+
+static u32 div64_round(u64 v, u32 d)
+{
+	return div_u64(v + (d >> 1), d);
+}
+
+static u32 div64_round_up(u64 v, u32 d)
+{
+	return div_u64(v + d - 1, d);
+}
+
+static int ar1335_code_to_bpp(struct ar1335_dev *sensor)
+{
+	switch (sensor->fmt.code) {
+	case MEDIA_BUS_FMT_SGRBG10_1X10:
+		return 10;
+	default:
+		return -EINVAL;
+	}
+}
+
+/* I2C register access */
+
+static int ar1335_write_regs(struct ar1335_dev *sensor, const __be16 *data,
+			     unsigned int count)
+{
+	struct i2c_client *client = sensor->i2c_client;
+	struct i2c_msg msg;
+	int ret;
+
+	msg.addr = client->addr;
+	msg.flags = client->flags;
+	msg.buf = (u8 *)data;
+	msg.len = count * sizeof(*data);
+
+	ret = i2c_transfer(client->adapter, &msg, 1);
+	if (ret < 0) {
+		dev_err(&client->dev, "%s: I2C write error\n", __func__);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int ar1335_write_reg(struct ar1335_dev *sensor, u16 reg, u16 val)
+{
+	__be16 buf[2] = { cpu_to_be16(reg), cpu_to_be16(val) };
+
+	return ar1335_write_regs(sensor, buf, 2);
+}
+
+static int ar1335_read_reg(struct ar1335_dev *sensor, u16 reg, u16 *val)
+{
+	struct i2c_client *client = sensor->i2c_client;
+	__be16 reg_addr = cpu_to_be16(reg);
+	__be16 reg_val;
+	struct i2c_msg msgs[2] = {
+		{
+			.addr = client->addr,
+			.flags = 0,
+			.len = sizeof(reg_addr),
+			.buf = (u8 *)&reg_addr,
+		},
+		{
+			.addr = client->addr,
+			.flags = I2C_M_RD,
+			.len = sizeof(reg_val),
+			.buf = (u8 *)&reg_val,
+		},
+	};
+	int ret;
+
+	ret = i2c_transfer(client->adapter, msgs, ARRAY_SIZE(msgs));
+	if (ret < 0) {
+		dev_err(&client->dev, "%s: I2C read error\n", __func__);
+		return ret;
+	}
+
+	*val = be16_to_cpu(reg_val);
+	return 0;
+}
+
+static int ar1335_write_reg_table(struct ar1335_dev *sensor,
+				  const struct ar1335_reg *regs,
+				  unsigned int count)
+{
+	unsigned int i;
+	int ret;
+
+	for (i = 0; i < count; i++) {
+		ret = ar1335_write_reg(sensor, regs[i].addr, regs[i].val);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+}
+
+/* Sensor configuration */
+
+static int ar1335_set_geometry(struct ar1335_dev *sensor)
+{
+	u16 out_width = sensor->fmt.width;
+	u16 out_height = sensor->fmt.height;
+	u16 x_start, y_start, x_end, y_end;
+	__be16 regs[9];
+
+	/* Center-crop the readout window to match the output size */
+	x_start = AR1335_MIN_X_ADDR_START +
+		  (AR1335_MAX_X_ADDR_END - AR1335_MIN_X_ADDR_START - out_width) / 2;
+	y_start = AR1335_MIN_Y_ADDR_START +
+		  (AR1335_MAX_Y_ADDR_END - AR1335_MIN_Y_ADDR_START - out_height) / 2;
+	x_end = x_start + out_width - 1;
+	y_end = y_start + out_height - 1;
+
+	regs[0] = cpu_to_be16(AR1335_REG_FRAME_LENGTH_LINES);
+	regs[1] = cpu_to_be16(out_height + sensor->ctrls.vblank->val);
+	regs[2] = cpu_to_be16(out_width + sensor->ctrls.hblank->val);
+	regs[3] = cpu_to_be16(x_start);
+	regs[4] = cpu_to_be16(y_start);
+	regs[5] = cpu_to_be16(x_end);
+	regs[6] = cpu_to_be16(y_end);
+	regs[7] = cpu_to_be16(out_width);
+	regs[8] = cpu_to_be16(out_height);
+
+	return ar1335_write_regs(sensor, regs, ARRAY_SIZE(regs));
+}
+
+static int ar1335_set_gains(struct ar1335_dev *sensor)
+{
+	int green = sensor->ctrls.gain->val;
+	int red = max(green + sensor->ctrls.red_balance->val, 0);
+	int blue = max(green + sensor->ctrls.blue_balance->val, 0);
+	unsigned int gain = min(red, min(green, blue));
+	unsigned int analog = min(gain, 64u);
+	__be16 regs[5];
+
+	red   = min(red   - analog + 64, 511u);
+	green = min(green - analog + 64, 511u);
+	blue  = min(blue  - analog + 64, 511u);
+	regs[0] = cpu_to_be16(AR1335_REG_GREEN1_GAIN);
+	regs[1] = cpu_to_be16(green << 7 | analog);
+	regs[2] = cpu_to_be16(blue  << 7 | analog);
+	regs[3] = cpu_to_be16(red   << 7 | analog);
+	regs[4] = cpu_to_be16(green << 7 | analog);
+
+	return ar1335_write_regs(sensor, regs, ARRAY_SIZE(regs));
+}
+
+static u32 calc_pll(struct ar1335_dev *sensor, u32 freq, u16 *pre_ptr,
+		    u16 *mult_ptr)
+{
+	u16 pre = 1, mult = 1, new_pre;
+	u32 pll = AR1335_PLL_MAX + 1;
+
+	for (new_pre = 1; new_pre < 64; new_pre++) {
+		u32 new_pll;
+		u32 new_mult = div64_round_up((u64)freq * new_pre,
+					      sensor->extclk_freq);
+
+		if (new_mult < 32)
+			continue;
+		if (new_mult > 254)
+			break;
+		if (sensor->extclk_freq * (u64)new_mult <
+		    AR1335_PLL_MIN * new_pre)
+			continue;
+		if (sensor->extclk_freq * (u64)new_mult >
+		    AR1335_PLL_MAX * new_pre)
+			break;
+		new_pll = div64_round_up(sensor->extclk_freq * (u64)new_mult,
+					 new_pre);
+		if (new_pll < pll) {
+			pll = new_pll;
+			pre = new_pre;
+			mult = new_mult;
+		}
+	}
+
+	pll = div64_round(sensor->extclk_freq * (u64)mult, pre);
+	*pre_ptr = pre;
+	*mult_ptr = mult;
+
+	return pll;
+}
+
+static void ar1335_calc_pll(struct ar1335_dev *sensor)
+{
+	unsigned int pixel_clock;
+	u16 pre, mult;
+	u32 vco;
+	int bpp;
+
+	pixel_clock = AR1335_PIXEL_CLOCK_RATE * 2 / sensor->lane_count;
+	bpp = ar1335_code_to_bpp(sensor);
+	if (bpp < 0)
+		return;
+
+	sensor->pll.vt_pix = bpp / 2;
+	vco = pixel_clock * sensor->pll.vt_pix;
+
+	calc_pll(sensor, vco, &pre, &mult);
+
+	sensor->pll.pre = sensor->pll.pre2 = pre;
+	sensor->pll.mult = sensor->pll.mult2 = mult;
+}
+
+static int ar1335_pll_config(struct ar1335_dev *sensor)
+{
+	__be16 pll_regs[] = {
+		cpu_to_be16(AR1335_REG_VT_PIX_CLK_DIV),
+		cpu_to_be16(sensor->pll.vt_pix),
+		cpu_to_be16(1),
+		cpu_to_be16((sensor->pll.pre2 << 8) | sensor->pll.pre),
+		cpu_to_be16((sensor->pll.mult2 << 8) | sensor->pll.mult),
+		cpu_to_be16(sensor->pll.vt_pix * 2),
+		cpu_to_be16(1),
+	};
+
+	ar1335_calc_pll(sensor);
+
+	return ar1335_write_regs(sensor, pll_regs, ARRAY_SIZE(pll_regs));
+}
+
+/* Streaming */
+
+static int ar1335_start_streaming(struct ar1335_dev *sensor)
+{
+	int ret;
+
+	ret = ar1335_write_reg(sensor, AR1335_REG_RESET,
+			       AR1335_REG_RESET_DEFAULTS);
+	if (ret)
+		return ret;
+
+	ret = ar1335_set_geometry(sensor);
+	if (ret)
+		return ret;
+
+	ret = ar1335_pll_config(sensor);
+	if (ret)
+		return ret;
+
+	ret = __v4l2_ctrl_handler_setup(&sensor->ctrls.handler);
+	if (ret)
+		return ret;
+
+	/* Exit LP-11 mode on clock and data lanes */
+	ret = ar1335_write_reg(sensor, AR1335_REG_HISPI_CONTROL_STATUS, 0);
+	if (ret)
+		return ret;
+
+	ret = ar1335_write_reg(sensor, AR1335_REG_RESET,
+			       AR1335_REG_RESET_DEFAULTS |
+			       AR1335_REG_RESET_STREAM);
+	if (ret)
+		return ret;
+
+	msleep(200);
+	return 0;
+}
+
+static int ar1335_stop_streaming(struct ar1335_dev *sensor)
+{
+	int ret;
+
+	/* Reset gain to avoid all-white pixels on next stream on */
+	ret = ar1335_write_reg(sensor, AR1335_REG_GLOBAL_GAIN, 0x2000);
+	if (ret)
+		return ret;
+
+	return ar1335_write_reg(sensor, AR1335_REG_RESET,
+				AR1335_REG_RESET_DEFAULTS);
+}
+
+/* V4L2 subdev operations */
+
+static void ar1335_adj_fmt(struct v4l2_mbus_framefmt *fmt)
+{
+	fmt->width = clamp(ALIGN(fmt->width, 4), AR1335_WIDTH_MIN,
+			   AR1335_WIDTH_MAX);
+	fmt->height = clamp(ALIGN(fmt->height, 4), AR1335_HEIGHT_MIN,
+			    AR1335_HEIGHT_MAX);
+	fmt->code = MEDIA_BUS_FMT_SGRBG10_1X10;
+	fmt->field = V4L2_FIELD_NONE;
+	fmt->colorspace = V4L2_COLORSPACE_SRGB;
+	fmt->ycbcr_enc = V4L2_YCBCR_ENC_DEFAULT;
+	fmt->quantization = V4L2_QUANTIZATION_FULL_RANGE;
+	fmt->xfer_func = V4L2_XFER_FUNC_DEFAULT;
+}
+
+static int ar1335_get_fmt(struct v4l2_subdev *sd,
+			  struct v4l2_subdev_state *state,
+			  struct v4l2_subdev_format *format)
+{
+	struct ar1335_dev *sensor = to_ar1335_dev(sd);
+
+	if (format->pad)
+		return -EINVAL;
+
+	mutex_lock(&sensor->lock);
+	format->format = sensor->fmt;
+	mutex_unlock(&sensor->lock);
+
+	return 0;
+}
+
+static int ar1335_set_fmt(struct v4l2_subdev *sd,
+			  struct v4l2_subdev_state *state,
+			  struct v4l2_subdev_format *format)
+{
+	struct ar1335_dev *sensor = to_ar1335_dev(sd);
+	struct v4l2_mbus_framefmt *fmt = &format->format;
+	int max_vblank, max_hblank, vblank, hblank;
+	int ret = 0;
+
+	ar1335_adj_fmt(fmt);
+
+	mutex_lock(&sensor->lock);
+
+	if (format->which == V4L2_SUBDEV_FORMAT_TRY) {
+		mutex_unlock(&sensor->lock);
+		return 0;
+	}
+
+	sensor->fmt = *fmt;
+	ar1335_calc_pll(sensor);
+
+	hblank = AR1335_WIDTH_BLANKING_MIN;
+	vblank = AR1335_HEIGHT_BLANKING_MIN;
+
+	max_hblank = AR1335_TOTAL_WIDTH_MAX - sensor->fmt.width;
+	ret = __v4l2_ctrl_modify_range(sensor->ctrls.hblank,
+				       AR1335_WIDTH_BLANKING_MIN,
+				       max_hblank, 1, hblank);
+	if (ret)
+		goto unlock;
+
+	ret = __v4l2_ctrl_s_ctrl(sensor->ctrls.hblank, hblank);
+	if (ret)
+		goto unlock;
+
+	max_vblank = AR1335_TOTAL_HEIGHT_MAX - sensor->fmt.height;
+	ret = __v4l2_ctrl_modify_range(sensor->ctrls.vblank,
+				       AR1335_HEIGHT_BLANKING_MIN,
+				       max_vblank, 2, vblank);
+	if (ret)
+		goto unlock;
+
+	ret = __v4l2_ctrl_s_ctrl(sensor->ctrls.vblank, vblank);
+	if (ret)
+		goto unlock;
+
+	ret = __v4l2_ctrl_modify_range(sensor->ctrls.exposure,
+				       sensor->ctrls.exposure->minimum,
+				       sensor->fmt.height + vblank - 4,
+				       sensor->ctrls.exposure->step,
+				       sensor->ctrls.exposure->default_value);
+
+unlock:
+	mutex_unlock(&sensor->lock);
+	return ret;
+}
+
+static const u16 ar1335_test_pattern_values[] = {
+	0x000,
+	0x001,
+	0x002,
+	0x003,
+	0x100,
+	0x101,
+};
+
+static const char * const ar1335_test_pattern_menu[] = {
+	"Normal pixel operation",
+	"Solid color",
+	"100% Color Bar",
+	"Fade-to-Gray Color Bars",
+	"Walking 1s (10-bit)",
+	"Walking 1s (8-bit)",
+};
+
+static int ar1335_s_ctrl(struct v4l2_ctrl *ctrl)
+{
+	struct v4l2_subdev *sd = ctrl_to_sd(ctrl);
+	struct ar1335_dev *sensor = to_ar1335_dev(sd);
+	int exp_max;
+	int ret;
+
+	switch (ctrl->id) {
+	case V4L2_CID_VBLANK:
+		exp_max = sensor->fmt.height + ctrl->val - 4;
+		__v4l2_ctrl_modify_range(sensor->ctrls.exposure,
+					 sensor->ctrls.exposure->minimum,
+					 exp_max,
+					 sensor->ctrls.exposure->step,
+					 sensor->ctrls.exposure->default_value);
+		ret = ar1335_set_geometry(sensor);
+		break;
+	case V4L2_CID_HBLANK:
+		ret = ar1335_set_geometry(sensor);
+		break;
+	case V4L2_CID_ANALOGUE_GAIN:
+		ret = ar1335_write_reg(sensor,
+				       AR1335_REG_ANA_GAIN_CODE_GLOBAL,
+				       ctrl->val);
+		break;
+	case V4L2_CID_GAIN:
+	case V4L2_CID_RED_BALANCE:
+	case V4L2_CID_BLUE_BALANCE:
+		ret = ar1335_set_gains(sensor);
+		break;
+	case V4L2_CID_EXPOSURE:
+		ret = ar1335_write_reg(sensor,
+				       AR1335_REG_COARSE_INTEGRATION_TIME,
+				       ctrl->val);
+		break;
+	case V4L2_CID_TEST_PATTERN:
+		ret = ar1335_write_reg(sensor, AR1335_REG_TEST_PATTERN_MODE,
+				       ar1335_test_pattern_values[ctrl->val]);
+		break;
+	default:
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+static const struct v4l2_ctrl_ops ar1335_ctrl_ops = {
+	.s_ctrl = ar1335_s_ctrl,
+};
+
+static int ar1335_init_controls(struct ar1335_dev *sensor)
+{
+	const struct v4l2_ctrl_ops *ops = &ar1335_ctrl_ops;
+	struct ar1335_ctrls *ctrls = &sensor->ctrls;
+	struct v4l2_ctrl_handler *hdl = &ctrls->handler;
+	struct v4l2_ctrl *link_freq;
+	int max_vblank, max_hblank;
+	int ret;
+
+	v4l2_ctrl_handler_init(hdl, 32);
+	hdl->lock = &sensor->lock;
+
+	v4l2_ctrl_new_std(hdl, ops, V4L2_CID_ANALOGUE_GAIN,
+			  AR1335_ANA_GAIN_MIN, AR1335_ANA_GAIN_MAX,
+			  AR1335_ANA_GAIN_STEP, AR1335_ANA_GAIN_DEFAULT);
+
+	ctrls->gain = v4l2_ctrl_new_std(hdl, ops, V4L2_CID_GAIN,
+					0, 511, 1, 40);
+	ctrls->red_balance = v4l2_ctrl_new_std(hdl, ops, V4L2_CID_RED_BALANCE,
+					       -512, 511, 1, 0);
+	ctrls->blue_balance = v4l2_ctrl_new_std(hdl, ops,
+						V4L2_CID_BLUE_BALANCE,
+						-512, 511, 1, 0);
+	v4l2_ctrl_cluster(3, &ctrls->gain);
+
+	max_hblank = AR1335_TOTAL_WIDTH_MAX - AR1335_WIDTH_MAX;
+	ctrls->hblank = v4l2_ctrl_new_std(hdl, ops, V4L2_CID_HBLANK,
+					  AR1335_WIDTH_BLANKING_MIN,
+					  max_hblank, 1,
+					  AR1335_WIDTH_BLANKING_MIN);
+
+	max_vblank = AR1335_TOTAL_HEIGHT_MAX - AR1335_HEIGHT_MAX;
+	ctrls->vblank = v4l2_ctrl_new_std(hdl, ops, V4L2_CID_VBLANK,
+					  AR1335_HEIGHT_BLANKING_MIN,
+					  max_vblank, 2,
+					  AR1335_HEIGHT_BLANKING_MIN);
+	v4l2_ctrl_cluster(2, &ctrls->hblank);
+
+	ctrls->pixrate = v4l2_ctrl_new_std(hdl, ops, V4L2_CID_PIXEL_RATE,
+					   AR1335_PIXEL_CLOCK_MIN,
+					   AR1335_PIXEL_CLOCK_MAX, 1,
+					   AR1335_PIXEL_CLOCK_RATE);
+
+	ctrls->exposure = v4l2_ctrl_new_std(hdl, ops, V4L2_CID_EXPOSURE,
+					    0, 65535, 1, 0x0C2E);
+
+	link_freq = v4l2_ctrl_new_int_menu(hdl, ops, V4L2_CID_LINK_FREQ,
+					   ARRAY_SIZE(ar1335_link_frequencies) - 1,
+					   0, ar1335_link_frequencies);
+	if (link_freq)
+		link_freq->flags |= V4L2_CTRL_FLAG_READ_ONLY;
+
+	ctrls->test_pattern =
+		v4l2_ctrl_new_std_menu_items(hdl, ops, V4L2_CID_TEST_PATTERN,
+					     ARRAY_SIZE(ar1335_test_pattern_menu) - 1,
+					     0, 0, ar1335_test_pattern_menu);
+
+	if (hdl->error) {
+		ret = hdl->error;
+		goto free_ctrls;
+	}
+
+	sensor->sd.ctrl_handler = hdl;
+	return 0;
+
+free_ctrls:
+	v4l2_ctrl_handler_free(hdl);
+	return ret;
+}
+
+/* Power management */
+
+static int ar1335_power_off(struct device *dev)
+{
+	struct v4l2_subdev *sd = dev_get_drvdata(dev);
+	struct ar1335_dev *sensor = to_ar1335_dev(sd);
+
+	clk_disable_unprepare(sensor->extclk);
+	gpiod_set_value_cansleep(sensor->reset_gpio, 1);
+	regulator_bulk_disable(AR1335_NUM_SUPPLIES, sensor->supplies);
+
+	return 0;
+}
+
+static int ar1335_power_on(struct device *dev)
+{
+	struct v4l2_subdev *sd = dev_get_drvdata(dev);
+	struct ar1335_dev *sensor = to_ar1335_dev(sd);
+	int ret;
+
+	ret = regulator_bulk_enable(AR1335_NUM_SUPPLIES, sensor->supplies);
+	if (ret)
+		return ret;
+
+	ret = clk_prepare_enable(sensor->extclk);
+	if (ret)
+		goto disable_regulators;
+	// Hold in reset for 1ms
+	gpiod_set_value_cansleep(sensor->reset_gpio, 1);
+	usleep_range(1000, 2000);
+	// Release from reset
+	gpiod_set_value_cansleep(sensor->reset_gpio, 0);
+	usleep_range(1000, 2000);
+
+	ret = ar1335_write_reg_table(sensor, ar1335_init_regs,
+				     ARRAY_SIZE(ar1335_init_regs));
+	if (ret)
+		goto assert_reset;
+
+	ret = ar1335_write_reg(sensor, AR1335_REG_SERIAL_FORMAT,
+			       AR1335_REG_SERIAL_FORMAT_MIPI |
+			       sensor->lane_count);
+	if (ret)
+		goto assert_reset;
+
+	ret = ar1335_write_reg(sensor, AR1335_REG_HISPI_TEST_MODE,
+			       ((0x40 << sensor->lane_count) - 0x40) |
+			       AR1335_REG_HISPI_TEST_MODE_LP11);
+	if (ret)
+		goto assert_reset;
+
+	ret = ar1335_write_reg(sensor, AR1335_REG_ROW_SPEED,
+			       0x110 | 4 / sensor->lane_count);
+	if (ret)
+		goto assert_reset;
+
+	return 0;
+
+assert_reset:
+	gpiod_set_value_cansleep(sensor->reset_gpio, 1);
+	clk_disable_unprepare(sensor->extclk);
+disable_regulators:
+	regulator_bulk_disable(AR1335_NUM_SUPPLIES, sensor->supplies);
+	return ret;
+}
+
+static const struct dev_pm_ops ar1335_pm_ops = {
+	SET_RUNTIME_PM_OPS(ar1335_power_off, ar1335_power_on, NULL)
+};
+
+/* V4L2 subdev operations */
+
+static int ar1335_enum_mbus_code(struct v4l2_subdev *sd,
+				 struct v4l2_subdev_state *state,
+				 struct v4l2_subdev_mbus_code_enum *code)
+{
+	if (code->index)
+		return -EINVAL;
+
+	code->code = MEDIA_BUS_FMT_SGRBG10_1X10;
+	return 0;
+}
+
+static int ar1335_enum_frame_size(struct v4l2_subdev *sd,
+				  struct v4l2_subdev_state *state,
+				  struct v4l2_subdev_frame_size_enum *fse)
+{
+	if (fse->index)
+		return -EINVAL;
+
+	if (fse->code != MEDIA_BUS_FMT_SGRBG10_1X10)
+		return -EINVAL;
+
+	fse->min_width = AR1335_WIDTH_MIN;
+	fse->max_width = AR1335_WIDTH_MAX;
+	fse->min_height = AR1335_HEIGHT_MIN;
+	fse->max_height = AR1335_HEIGHT_MAX;
+
+	return 0;
+}
+
+static int ar1335_s_stream(struct v4l2_subdev *sd, int enable)
+{
+	struct ar1335_dev *sensor = to_ar1335_dev(sd);
+	struct device *dev = &sensor->i2c_client->dev;
+	int ret;
+
+	mutex_lock(&sensor->lock);
+
+	if (enable) {
+		ret = pm_runtime_resume_and_get(dev);
+		if (ret < 0)
+			goto unlock;
+
+		ret = ar1335_start_streaming(sensor);
+		if (ret) {
+			pm_runtime_put_sync(dev);
+			goto unlock;
+		}
+	} else {
+		ret = ar1335_stop_streaming(sensor);
+		pm_runtime_mark_last_busy(dev);
+		pm_runtime_put_autosuspend(dev);
+	}
+
+unlock:
+	mutex_unlock(&sensor->lock);
+	return ret;
+}
+
+static const struct media_entity_operations ar1335_media_ops = {
+	.link_validate = v4l2_subdev_link_validate,
+};
+
+static const struct v4l2_subdev_core_ops ar1335_core_ops = {
+	.log_status = v4l2_ctrl_subdev_log_status,
+};
+
+static const struct v4l2_subdev_video_ops ar1335_video_ops = {
+	.s_stream = ar1335_s_stream,
+};
+
+static const struct v4l2_subdev_pad_ops ar1335_pad_ops = {
+	.enum_mbus_code = ar1335_enum_mbus_code,
+	.enum_frame_size = ar1335_enum_frame_size,
+	.get_fmt = ar1335_get_fmt,
+	.set_fmt = ar1335_set_fmt,
+};
+
+static const struct v4l2_subdev_ops ar1335_subdev_ops = {
+	.core = &ar1335_core_ops,
+	.video = &ar1335_video_ops,
+	.pad = &ar1335_pad_ops,
+};
+
+/* Chip identification */
+
+static int ar1335_identify_module(struct ar1335_dev *sensor)
+{
+	struct device *dev = &sensor->i2c_client->dev;
+	u16 chip_id;
+	int ret;
+
+	ret = ar1335_read_reg(sensor, AR1335_REG_CHIP_ID, &chip_id);
+	if (ret) {
+		dev_err(dev, "Failed to read chip ID\n");
+		return ret;
+	}
+
+	if (chip_id != AR1335_CHIP_ID) {
+		dev_err(dev, "Chip ID mismatch: expected 0x%04x, got 0x%04x\n",
+			AR1335_CHIP_ID, chip_id);
+		return -ENODEV;
+	}
+
+	dev_dbg(dev, "AR1335 chip ID: 0x%04x\n", chip_id);
+	return 0;
+}
+
+/* Probe and remove */
+
+static int ar1335_get_regulators(struct ar1335_dev *sensor)
+{
+	unsigned int i;
+
+	for (i = 0; i < AR1335_NUM_SUPPLIES; i++)
+		sensor->supplies[i].supply = ar1335_supply_names[i];
+
+	return devm_regulator_bulk_get(&sensor->i2c_client->dev,
+				       AR1335_NUM_SUPPLIES, sensor->supplies);
+}
+
+static int ar1335_probe(struct i2c_client *client)
+{
+	struct v4l2_fwnode_endpoint ep = {
+		.bus_type = V4L2_MBUS_CSI2_DPHY
+	};
+	struct device *dev = &client->dev;
+	struct fwnode_handle *endpoint;
+	struct ar1335_dev *sensor;
+	int ret;
+	dev_err(dev, "Probing AR1335 sensor\n");
+	sensor = devm_kzalloc(dev, sizeof(*sensor), GFP_KERNEL);
+	if (!sensor)
+		return -ENOMEM;
+
+	sensor->i2c_client = client;
+
+	v4l2_i2c_subdev_init(&sensor->sd, client, &ar1335_subdev_ops);
+
+	/* Parse endpoint configuration */
+	endpoint = fwnode_graph_get_endpoint_by_id(dev_fwnode(dev), 0, 0,
+						   FWNODE_GRAPH_ENDPOINT_NEXT);
+	if (!endpoint)
+		return dev_err_probe(dev, -EINVAL, "Endpoint node not found\n");
+
+	ret = v4l2_fwnode_endpoint_parse(endpoint, &ep);
+	fwnode_handle_put(endpoint);
+	if (ret)
+		return dev_err_probe(dev, ret, "Could not parse endpoint\n");
+
+	if (ep.bus_type != V4L2_MBUS_CSI2_DPHY)
+		return dev_err_probe(dev, -EINVAL,
+				     "Invalid bus type, must be MIPI CSI2\n");
+
+	sensor->lane_count = ep.bus.mipi_csi2.num_data_lanes;
+	switch (sensor->lane_count) {
+	case 1:
+	case 2:
+	case 4:
+		break;
+	default:
+		return dev_err_probe(dev, -EINVAL,
+				     "Invalid number of MIPI data lanes\n");
+	}
+
+	/* Get clock and regulators */
+	sensor->extclk = devm_clk_get(dev, "extclk");
+	if (IS_ERR(sensor->extclk))
+		return dev_err_probe(dev, PTR_ERR(sensor->extclk),
+				     "Failed to get extclk\n");
+
+	sensor->extclk_freq = clk_get_rate(sensor->extclk);
+	if (sensor->extclk_freq < AR1335_EXTCLK_MIN ||
+	    sensor->extclk_freq > AR1335_EXTCLK_MAX)
+		return dev_err_probe(dev, -EINVAL,
+				     "Extclk frequency out of range: %u Hz\n",
+				     sensor->extclk_freq);
+	ret = ar1335_get_regulators(sensor);
+	if (ret)
+		return dev_err_probe(dev, ret, "Failed to get regulators\n");
+
+	sensor->reset_gpio = devm_gpiod_get_optional(dev, "reset",
+						      GPIOD_OUT_HIGH);
+	if (IS_ERR(sensor->reset_gpio))
+		return dev_err_probe(dev, PTR_ERR(sensor->reset_gpio),
+				     "Failed to get reset GPIO\n");
+
+	/* Power on and identify the sensor */
+	ret = ar1335_power_on(dev);
+	if (ret)
+		return dev_err_probe(dev, ret, "Failed to power on sensor\n");
+
+	ret = ar1335_identify_module(sensor);
+	if (ret)
+		goto error_power_off;
+	dev_info(dev, "Detected ON Semiconductor AR1335 sensor\n");
+
+	/* Initialize controls and subdev */
+	mutex_init(&sensor->lock);
+
+	ret = ar1335_init_controls(sensor);
+	if (ret)
+		goto error_mutex;
+
+	ar1335_adj_fmt(&sensor->fmt);
+
+	sensor->sd.flags |= V4L2_SUBDEV_FL_HAS_DEVNODE;
+	sensor->pad.flags = MEDIA_PAD_FL_SOURCE;
+	sensor->sd.entity.function = MEDIA_ENT_F_CAM_SENSOR;
+	sensor->sd.entity.ops = &ar1335_media_ops;
+
+	ret = media_entity_pads_init(&sensor->sd.entity, 1, &sensor->pad);
+	if (ret)
+		goto error_handler_free;
+
+	ret = v4l2_async_register_subdev_sensor(&sensor->sd);
+	if (ret)
+		goto error_media_entity;
+
+	/* Enable runtime PM */
+	pm_runtime_set_active(dev);
+	pm_runtime_enable(dev);
+	pm_runtime_idle(dev);
+	pm_runtime_set_autosuspend_delay(dev, 1000);
+	pm_runtime_use_autosuspend(dev);
+
+	return 0;
+
+error_media_entity:
+	media_entity_cleanup(&sensor->sd.entity);
+error_handler_free:
+	v4l2_ctrl_handler_free(&sensor->ctrls.handler);
+error_mutex:
+	mutex_destroy(&sensor->lock);
+error_power_off:
+	ar1335_power_off(dev);
+	return ret;
+}
+
+static void ar1335_remove(struct i2c_client *client)
+{
+	struct v4l2_subdev *sd = i2c_get_clientdata(client);
+	struct ar1335_dev *sensor = to_ar1335_dev(sd);
+
+	v4l2_async_unregister_subdev(&sensor->sd);
+	media_entity_cleanup(&sensor->sd.entity);
+	v4l2_ctrl_handler_free(&sensor->ctrls.handler);
+
+	pm_runtime_disable(&client->dev);
+	pm_runtime_set_suspended(&client->dev);
+
+	mutex_destroy(&sensor->lock);
+}
+
+static const struct of_device_id ar1335_id[] = {
+	{ .compatible = "onnn,ar1335" },
+	{ }
+};
+MODULE_DEVICE_TABLE(of, ar1335_id);
+
+static struct i2c_driver ar1335_driver = {
+	.driver = {
+		.name  = AR1335_NAME,
+		.of_match_table = ar1335_id,
+		.pm = &ar1335_pm_ops,
+	},
+	.probe = ar1335_probe,
+	.remove = ar1335_remove,
+};
+
+module_i2c_driver(ar1335_driver);
+
+MODULE_AUTHOR("William Bright <William.Bright@imd-tec.com>");
+MODULE_DESCRIPTION("ON Semiconductor AR1335 CMOS Image Sensor driver");
+MODULE_LICENSE("GPL v2");
-- 
2.43.0

